### ISSUE TITLE
🐛 Performance issue for production server

### ISSUE THREAD

**Issue by jimmidier** [SSWConsulting/SSW.SophieBot#483]
Cc: @pierssinclairssw @adamcogan 

Hi @ravidhaduk 

### Describe the Bug

As per our conversation, SophieBot on prod is working very slowly currently. It takes around 20 seconds to get response for each command.

### Expected Behavior

It should response after approx 5 seconds, as it used to be.

### Tasks
- [x] Investigate
- [x] Fix

### Screenshots

![image](https://user-images.githubusercontent.com/16027480/160338646-769e3642-abb2-4c38-959e-73ccccb0d016.png)
**Figure: Server response time has increased much**

### Definition of Done
- [x] All tasks are updated and closed
- [x] No high priority defects/bugs are on that user story
- [x] Code reviewed by one other team member (e.g. Checked by Bill)
- [x] All acceptance criteria have been met
- [x] Deployed to Dev
- [x] (If implemented) Pass automated UI testing
- [x] (If not implemented above) Do a manual test on all supported commands (at least one for each LUIS intent) against Dev
- [x] Complex code is documented (removing technical debt)
- [x] Deployed to Production
- [x] Add description about the changes for the PBI
- [x] Tick off the definition of done tasks in the PBI
- [x] (Optional) A Done Video, particularly for PBI with multiple stages

Thanks!

**Comment by ravidhaduk** [comment]
I have made a change to improve this and the response time has gone from 12 sec to 4.5 sec.
To make this faster, I changed the data retrival to filter appointments before clients which returns a much smaller data sets than doing it vice versa.

<img width="334" alt="Screen Shot 2022-04-08 at 4 33 26 pm" src="https://user-images.githubusercontent.com/80997507/162378022-2de76137-94bf-4e63-937d-51382dac8f3f.png">
Figure: New response time is 4.5 sec.

Since this is now as fast as it used to be, I will be closing this issue.

**Comment by jimmidier** [comment]
Reopening as it's still not resolved on production

**Comment by duanxinhuan** [comment]
As per conversaiton, the tasks are listed below:
 
- [ ] Investigate if we already record any performance numbers?
- [ ] Investigate the native way to integrate LUIS to Application Insights, log both intent and query.
- [ ] Integrate People API and log the response time for different endpoints.
- [ ] Analyse why Ravi’s change - [07eae9154a660c448747bdc7ed9da4bfc8a49923](https://github.com/SSWConsulting/SSW.SophieBot/commit/07eae9154a660c448747bdc7ed9da4bfc8a49923), slow down - [ ] the SophieBot.
- [ ] Add the annotation to Application Insights on DevOps pipeline when developer pushes the code.

**Comment by ravidhaduk** [comment]
we are investigating the performance issue of the SophieBot.
 
We have tried different ways to isolate the cause of the performance issue such as
Added telemetry at the few intents flows to measure where and why it is taking more time to respond. (Found out there’s a delay when sending a message to Teams)
Debugged the project locally and monitored each trace. (No issue found)
Ran local project on MS Teams to check whether the problem is at Teams end or not. (It was responding way faster than prod)
 
But couldn’t find any issue with the bot application code. Instead, everything points towards the infrastructure issue.
We have noticed the CPU and memory utilization spiking to 100% for a single query, and it seems like it is locking up the server and as a result, it is unable to process the responses that need to go to the bot channel.
 
So, what we did is we scaled up the App Service to P1V2 on dev, and suddenly latency has decreased from 44.2 to 12.8 seconds.
 
Following are the App Insights shots from Prod (S1 Tier) and Dev (P1V2) service of the query Who knows Angular?

![image](https://user-images.githubusercontent.com/80997507/172737134-a5c64840-05e7-4f03-b48a-daa70b14faf2.png)    
**Figure: SophieBot Prod took 44.2 seconds to complete the query**

![image](https://user-images.githubusercontent.com/80997507/172737155-b63e9b07-fded-4df5-a138-1d150cf336ed.png)
**Figure: SophieBot Dev took 12.8 seconds to complete the query**

In summary, we are confident that scaling up the service to P1V2 will provide a better bot experience. 
 
However, the cost will increase from $126.29 to $252.28 AUD/month.

As per the conversation with Patrick, Jim, and William L, we were using Windows server on Azure. If we use Linux server, the cost will be reduced from $252.28 to $111.25 for P1V2 tier.

- [ ] Create a new resource group calles SSW.Sophie.Bot.Tes and use Linux server and P1V2 tier.
- [ ] Deploy the app in this RB.
- [ ] Test the app and measure the performance. 
- [ ] If everything works as expected on Linux server, Switch our SophieBot Prod and Dev RG to Linux server and scaled up those with P1V2 tier.

**Comment by ravidhaduk** [comment]
As per my conversation with Kiki, he suggested me to use my Azure Credit to create a Test resource group instead create. it on SSW Primary account. This is because this group is for short time and it will help me to get familiar with Azure.

Here's the Test RG:  https://portal.azure.com/#@sswcom.onmicrosoft.com/resource/subscriptions/7f6a3d28-506e-4a8b-925d-854559fdb150/resourceGroups/SSW.Sophie.Bot.Test/overview

**Comment by ravidhaduk** [comment]
After deploying the bot to azure, I have been testing the bot on Microsoft Teams. It was working way faster than before.

Here's the server response time when I tested all the queries. 
I have asked more than 30 queries from 10.34AM to 10.40AM.
![image](https://user-images.githubusercontent.com/80997507/172740950-ea487efb-e982-4e2f-aaff-58b47a13955a.png)
**Figure: Average Server response time was 1.72 Seconds** 

Here's the App Insights of the query which was taking logest time to response. 
Query: **Who knows Angular?** and then **Show All** click.

![image](https://user-images.githubusercontent.com/80997507/172741758-77c7b5d8-e1bc-4a99-a75e-586ff3f81899.png)
**Figure: SophieBot Test took 11.2 seconds to complete the query**

**Comment by ravidhaduk** [comment]
As per the conversation with Jim, Patrick, William L, we’ve got to migrate SophieBot Azure service from Windows to Linux for SSW Dev and Prod.

- [ ] Create a new App Service Plan in SSWSophieBotDev resource group with Linux.
- [ ] Re-run the [pipeline](https://github.com/SSWConsulting/SSW.SophieBot/actions/workflows/SSW.SophieBot.dev.yml) of dev targeting Linux.
- [ ] Change the bot app's endpoint on Bot Service configuration to point to the Linux app, and note down the downtime
- [ ] Test the app.
- [ ] If it succeeds , remove the windows App Service Plan from SSWSophieBotDev.
- [ ] Follow the same method to switch Prod SSWSophieBot App Service.

**Comment by ravidhaduk** [comment]
Done - I have successfully switched Dev and Prod App Service from Windows to Linux.

I have tested all the supported queries in SophieBot. It was working as it is expected and much faster than before.

Since we have scaled up the pricing tier to P1V2, it really taking less time to respond than before.

![image](https://user-images.githubusercontent.com/80997507/172965540-78af6076-5adb-4bcc-b89e-f7dad7652bd2.png)
**Figure: Average response time was 3.85 seconds.**

**Comment by pierssinclairssw** [comment]
Hi All,

This appears to still be an issue. Camilla and Adriana will provide more info.

**Comment by AdriT9** [comment]
Hi Guys, 
The first answer to the query “Who works in Sydney office?" only appeared after 3 minutes and 9 seconds. After 16 seconds, the notification “unable to reach app. Please try again” popped up. I clicked “try again”, and a few minutes later Sophie responded. I then picked Steven Qiang and it took 10 seconds for his card to show up. All the queries afterwards took less than 8 seconds.

**Comment by jimmidier** [comment]
Might be related to #567. The team will keep watch for similar performance drop

**Comment by jimmidier** [comment]
Closing issue as per #567
